sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,224 @@
+  return true;
+}
+
+void BatchLogProcessor::DoBackgroundWork()
+{
+  auto timeout = schedule_delay_millis_;
+
+  while (true)
+  {
+    // Wait for `timeout` milliseconds
+    std::unique_lock<std::mutex> lk(cv_m_);
+    cv_.wait_for(lk, timeout);
+
+    if (is_shutdown_.load() == true)
+    {
+      DrainQueue();
+      return;
+    }
+
+    bool was_force_flush_called = is_force_flush_.load();
+
+    // Check if this export was the result of a force flush.
+    if (was_force_flush_called == true)
+    {
+      // Since this export was the result of a force flush, signal the
+      // main thread that the worker thread has been notified
+      is_force_flush_ = false;
+    }
+    else
+    {
+      // If the buffer was empty during the entire `timeout` time interval,
+      // go back to waiting. If this was a spurious wake-up, we export only if
+      // `buffer_` is not empty. This is acceptable because batching is a best
+      // mechanism effort here.
+      if (buffer_.empty() == true)
+      {
+        continue;
+      }
+    }
+
+    auto start = std::chrono::steady_clock::now();
+
+    Export(was_force_flush_called);
+
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    // Subtract the duration of this export call from the next `timeout`.
+    timeout = schedule_delay_millis_ - duration;
+  }
+}
+
+void BatchLogProcessor::Export(const bool was_force_flush_called)
+{
+  std::vector<std::unique_ptr<LogRecord>> records_arr;
+
+  size_t num_records_to_export;
+
+  if (was_force_flush_called == true)
+  {
+    num_records_to_export = buffer_.size();
+  }
+  else
+  {
+    num_records_to_export =
+        buffer_.size() >= max_export_batch_size_ ? max_export_batch_size_ : buffer_.size();
+  }
+
+  buffer_.Consume(
+      num_records_to_export, [&](CircularBufferRange<AtomicUniquePtr<LogRecord>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<LogRecord> &ptr) {
+          std::unique_ptr<LogRecord> swap_ptr = std::unique_ptr<LogRecord>(nullptr);
+          ptr.Swap(swap_ptr);
+          records_arr.push_back(std::unique_ptr<LogRecord>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  ExportResult export_status = exporter_->Export(records_arr);
+  if (export_status != ExportResult::kSuccess)
+  {
+    // Error
+  }
+
+  // Notify the main thread in case this export was the result of a force flush.
+  if (was_force_flush_called == true)
+  {
+    is_force_flush_notified_ = true;
+    while (is_force_flush_notified_.load() == true)
+    {
+      force_flush_cv_.notify_one();
+    }
+  }
+}
+
+void BatchLogProcessor::DrainQueue()
+{
+  while (buffer_.empty() == false)
+  {
+    Export(false);
+  }
+}
+
+// Note: Timeout functionality is currently not implemented
+bool BatchLogProcessor::Shutdown(std::chrono::microseconds timeout) noexcept
+{
+  if (is_shutdown_.load() == true)
```

Comment:
This is not atomic :) it would break. You just can't do it like this, no :)

You have to use [atomic exchange](https://en.cppreference.com/w/cpp/atomic/atomic/exchange).

Let me elaborate, as this is fundamental issue that might bite us hard.

- line 195 you do non-atomic compare
- line 200 you do non-atomic assignment
- imagine for a second that two threads A and B both called into `Shutdown` at the same time
- on multicore system you have zero guarantees about how the two threads are getting scheduled: you may race A and B executing A:195, B:195 - _at nearly the same time_ seeing `false` ; then threads A and B both assigning `is_shudown_` true, and both stuck.

You have to use atomic exchange to address this.

Reply:
Hmm I wasn't sure how the atomic swap-and-exchange would work here, since the `is_shutdown` variable would never need to be set to `false` again if it were already set to `true`. 

I added a lock instead that persists for the scope of this function. This will ensure the function doesn't get executed by more than 1 thread at once. 

Also, line 200, that used the assignment `operator =` originally, was in fact atomic (see[cpp ref](https://en.cppreference.com/w/cpp/atomic/atomic/operator=)), but I changed it to its equivalent `.store(true)` for clarity as well :)

Reply:
Also related is - I think this is a problem for traces as well. I can either file an issue for it or use the same implementation for traces as well in a later PR

Reply:
1. Maybe I'm misunderstanding the usage of the exchange operation- but I am not sure how the atomic swap-and-exchange/exchange would work here, since we would never want to mistakenly set the `is_shutdown` variable to `false` again if it were already set to `true`? 
* Instead, I added a lock that persists for the scope of this function, along with a global mutex. This will ensure the function doesn't get executed by more than 1 thread at once. Let me know if this works!

2. Also, in line 200 (now line 202): According to [cpp ref](https://en.cppreference.com/w/cpp/atomic/atomic/operator=), the assignment `operator =` is "Equivalent to `.store(...)`"  (but nonetheless, I've updated all occurences of it for clarity)

Reply:
Since this implementation is based off traces, let me know whether I should file an issue for it as well (or address it in a later PR if the changes are the same as the ones here)

Reply:
You'd do this:

```cpp
if (!is_shutdown._exchange(true))
{
...
}
```

How this code works:
- it atomically reads and changes the value, guaranteed once, safe if multiple threads are racing thru the same code concurrently
- if your thread is lucky to toggle the state to `true`, it'll be the only one executing the shutdown code
- next thread would try to assign `true` again, but the value is already true - so assignment works, however the function returns that the previous value has already been `true`, so the code inside the scope won't get executed.

It is not about `.load()` and `.store()` being non-atomic: both of them are atomic. But they are separate operations. You can't use them separately, two of them are already non-atomic :) ... You'd use `exchange()` instead.

Reply:
Ohh makes sense, makes sense - I was looking at compare and swap instead of this get and set exchange, so got super confused. Thanks for explaining!